File: app/lib/domain/entities/surah_difficulty.dart
import 'package:tilawa/domain/entities/surah_difficulty_table.dart';

/// How fast a surah fades, relative to an average one.
///
/// Three layers, each overriding the one before:
///
/// 1. **Measured.** Every surah is scored from the text itself — bulk, mean
///    verse length, and the share of five-word runs that recur elsewhere in the
///    Quran, which is the *mutashabihat* that make a hafiz lose their place.
///    `tools/build_surah_difficulty.py` does the measuring and writes
///    [SurahDifficultyTable]; the evidence for each surah is in a comment on
///    its row.
/// 2. **Curated.** A few surahs stand apart for reasons the text cannot show.
///    Al-Fatiha is recited in every unit of every prayer; Al-Kahf on Fridays,
///    Al-Mulk before sleep, Yaseen and the short surahs of Juz Amma are all
///    rehearsed constantly without anyone calling it revision. Measurement puts
///    Al-Kahf, Yaseen and Al-Mulk in the middle tier, which is right about the
///    text and wrong about the person.
/// 3. **Personal.** What the reciter says they actually recite beats any
///    assumption about what people in general recite.
///
/// The coefficient divides the half-life: above 1 the surah halves sooner and
/// comes round for revision more often; below 1 it holds longer.
///
/// No curated "hard" list is kept. Measurement reproduces the four surahs that
/// used to be listed by hand — Al-Baqarah, An-Nisa, Al-Ma'idah and Al-An'am all
/// land in the top tier on their own — so listing them again would be data
/// duplicated in two places. A test asserts it instead, which is where a claim
/// like that belongs.
///
/// The API applies the same rules in `SurahDifficulty.cs`. The two must stay in
/// step — change one, change the other, and update both test suites.
class SurahDifficulty {
  const SurahDifficulty._();

  /// Tier coefficients, from the surah that holds longest to the one that
  /// fades fastest. Indexed by [SurahDifficultyTable.tierBySurah].
  static const List<double> tierCoefficients = [0.5, 0.75, 1.0, 1.25, 1.5];

  /// The tier a surah falls in when nothing is known about it.
  static const int defaultTier = 2;

  static const double easiestCoefficient = 0.5;
  static const double neutralCoefficient = 1.0;
  static const double hardestCoefficient = 1.5;

  /// Surahs carried by habit rather than by revision.
  static const Set<int> frequentlyRecitedSurahs = {
    1, // Al-Fatiha — every rak'ah of every prayer
    18, // Al-Kahf — Fridays
    36, // Yaseen
    67, // Al-Mulk — before sleep
  };

  /// Juz Amma, recited in daily prayer.
  static const int juzAmmaFirstSurah = 78;
  static const int juzAmmaLastSurah = 114;

  /// The tier this surah's text alone puts it in.
  static int measuredTier(int surahNumber) =>
      SurahDifficultyTable.tierBySurah[surahNumber] ?? defaultTier;

  /// What the text alone says, before any habit is taken into account.
  static double measuredCoefficient(int surahNumber) =>
      tierCoefficients[measuredTier(surahNumber)];

  /// Whether this surah is one the population at large rehearses anyway.
  static bool isHabituallyRecited(int surahNumber) =>
      frequentlyRecitedSurahs.contains(surahNumber) ||
      (surahNumber >= juzAmmaFirstSurah && surahNumber <= juzAmmaLastSurah);

  /// The multiplier applied to this surah's decay rate.
  ///
  /// Once the reciter has told us what they actually recite, that replaces the
  /// assumption about what people in general recite — including for surahs on
  /// the curated list. Someone who never opens Al-Kahf on a Friday should not
  /// have the app quietly assume they do.
  static double coefficientFor(
    int surahNumber, {
    Set<int>? personalFrequentlyRecited,
  }) {
    if (personalFrequentlyRecited != null) {
      return personalFrequentlyRecited.contains(surahNumber)
          ? easiestCoefficient
          : measuredCoefficient(surahNumber);
    }
    return isHabituallyRecited(surahNumber)
        ? easiestCoefficient
        : measuredCoefficient(surahNumber);
  }

  /// Whether this surah decays faster than average.
  static bool isHard(int surahNumber) =>
      coefficientFor(surahNumber) > neutralCoefficient;

  /// Whether this surah holds longer than average.
  static bool isFrequentlyRecited(int surahNumber) =>
      coefficientFor(surahNumber) < neutralCoefficient;
}

File: backend/Tilawa.Api/Services/SurahDifficulty.cs
namespace Tilawa.Api.Services;

/// <summary>
/// How fast a surah fades, relative to an average one.
/// </summary>
/// <remarks>
/// <para>
/// Three layers, each overriding the one before. <b>Measured:</b> every surah is
/// scored from the text — bulk, mean verse length, and the share of five-word
/// runs that recur elsewhere in the Quran, which is the <em>mutashabihat</em>
/// that make a hafiz lose their place. <c>tools/build_surah_difficulty.py</c>
/// does the measuring and writes <see cref="SurahDifficultyTable"/>.
/// <b>Curated:</b> a few surahs stand apart for reasons the text cannot show —
/// Al-Fatiha is in every rak'ah, Al-Kahf on Fridays, Al-Mulk before sleep,
/// Yaseen and Juz Amma constantly. <b>Personal:</b> what the reciter says they
/// recite beats any assumption about people in general.
/// </para>
/// <para>
/// The coefficient divides the half-life: above 1 the surah halves sooner;
/// below 1 it holds longer.
/// </para>
/// <para>
/// No curated "hard" list is kept. Measurement reproduces the four surahs that
/// used to be listed by hand — Al-Baqarah, An-Nisa, Al-Ma'idah and Al-An'am all
/// reach the top tier on their own. A test asserts it rather than duplicating
/// the data.
/// </para>
/// <para>
/// The app applies the same rules in <c>surah_difficulty.dart</c>. The two must
/// stay in step — change one, change the other, and update both test suites.
/// </para>
/// </remarks>
public static class SurahDifficulty
{
    /// <summary>
    /// Tier coefficients, from the surah that holds longest to the one that
    /// fades fastest. Indexed by <see cref="SurahDifficultyTable.TierBySurah"/>.
    /// </summary>
    public static readonly IReadOnlyList<double> TierCoefficients =
        new[] { 0.5, 0.75, 1.0, 1.25, 1.5 };

    /// <summary>The tier a surah falls in when nothing is known about it.</summary>
    public const int DefaultTier = 2;

    public const double EasiestCoefficient = 0.5;
    public const double NeutralCoefficient = 1.0;
    public const double HardestCoefficient = 1.5;

    public const int JuzAmmaFirstSurah = 78;
    public const int JuzAmmaLastSurah = 114;

    /// <summary>Surahs carried by habit rather than by revision.</summary>
    public static readonly IReadOnlySet<int> FrequentlyRecitedSurahs = new HashSet<int>
    {
        1,  // Al-Fatiha — every rak'ah of every prayer
        18, // Al-Kahf — Fridays
        36, // Yaseen
        67, // Al-Mulk — before sleep
    };

    /// <summary>The tier this surah's text alone puts it in.</summary>
    public static int MeasuredTier(int surahNumber) =>
        SurahDifficultyTable.TierBySurah.TryGetValue(surahNumber, out var tier)
            ? tier
            : DefaultTier;

    /// <summary>What the text alone says, before any habit is considered.</summary>
    public static double MeasuredCoefficient(int surahNumber) =>
        TierCoefficients[MeasuredTier(surahNumber)];

    /// <summary>Whether the population at large rehearses this surah anyway.</summary>
    public static bool IsHabituallyRecited(int surahNumber) =>
        FrequentlyRecitedSurahs.Contains(surahNumber) ||
        (surahNumber >= JuzAmmaFirstSurah && surahNumber <= JuzAmmaLastSurah);

    /// <summary>
    /// The multiplier applied to this surah's decay rate. Once the reciter has
    /// told us what they actually recite, that replaces the assumption about
    /// what people in general recite — including for the curated list.
    /// </summary>
    public static double CoefficientFor(
        int surahNumber,
        IReadOnlySet<int>? personalFrequentlyRecited = null)
    {
        if (personalFrequentlyRecited is not null)
        {
            return personalFrequentlyRecited.Contains(surahNumber)
                ? EasiestCoefficient
                : MeasuredCoefficient(surahNumber);
        }

        return IsHabituallyRecited(surahNumber)
            ? EasiestCoefficient
            : MeasuredCoefficient(surahNumber);
    }

    /// <summary>Whether this surah decays faster than average.</summary>
    public static bool IsHard(int surahNumber) =>
        CoefficientFor(surahNumber) > NeutralCoefficient;

    /// <summary>Whether this surah holds longer than average.</summary>
    public static bool IsFrequentlyRecited(int surahNumber) =>
        CoefficientFor(surahNumber) < NeutralCoefficient;
}
